Docs refactor composables page (#5996)
vuestorefront/vue-storefront: packages/core/docs/guide/error-handling.md

# Error handling

Each [composable](./composables) has a computed property called `errors` which stores errors from all methods within that composable. It's an object with keys matching the names of the methods within the composable, and Error instance or `null` as a value.

```vue
<template>
  <button @click="addItem(product)">Add to cart</button>
  <div v-if="error.addItem">{{ error.addItem.message }}</div>
</template>

<script>
export default {
  setup() {
    const { addItem, error } = useCart();

    return {
      addItem,
      error
    };
  }
};
</script>
```
The `@vue-storefront/core` package exports interfaces for `error` objects for every composable. Let's take a look at the one from the `useCart` composable:

```ts
export interface UseCartErrors {
  addItem: Error;
  removeItem: Error;
  updateItemQty: Error;
  load: Error;
  clear: Error;
  applyCoupon: Error;
  removeCoupon: Error;
}
```

:::details Where does the error come from?

Inside each async method, we catch thrown errors and save them in the `error` object under the key corresponding to the called method.

```ts
const { addItem, error } = useCart();

addItem({ product: null }); // throws an error

> error.value.addItem; // here you have error raised by addItem function

Reviewer: ```suggestion
error.value.addItem; // here you access error thrown by addItem function
```

```

:::

:::details Where can I find an interface of the error property for a specific composable?

When you are writing a code inside a script part of the Vue component, your IDE should give you hints on every different type of the composable. That's why you probably do not need to check these interfaces in the core's code.

> However, if you still want to check interfaces, you could find them inside [`packages/core/core/src/types.ts`](https://github.com/vuestorefront/vue-storefront/blob/next/packages/core/core/src/types.ts).

Reviewer: ```suggestion
However, if you still want to check interfaces, you could find them in [core API reference](../core/api-reference/core).
```

Please test this link

# Error handling

A flexible way of error handling is essential for a framework like Vue Storefront. We decided to put all errors in reactive objects so they can be easily managed in the template.

Each composable returns an `error` computed property. It is an object which has names of async functions from composable as keys and Error instance or `null` as a value.

```vue
<template>
  <button @click="addItem(product)">Add to cart</button>
  <div v-if="error.addItem">{{ error.addItem.message }}</div>
</template>

<script>
export default {
  setup() {
    const { addItem, error } = useCart();

    return {
      addItem,
      error
    };
  }
};
</script>
```

> There is a dedicated TypeScript interface for every composable. Take a look at this one from `useCart`:

Reviewer: ```suggestion
The `@vue-storefront/core` package exports interfaces for `error` objects for every composable. Let's take a look at the one from the `useCart` composable:
```

```ts
export interface UseCartErrors {
  addItem: Error;
  removeItem: Error;
  updateItemQty: Error;
  load: Error;
  clear: Error;
  applyCoupon: Error;
  removeCoupon: Error;
}
```

:::details Where does the error come from?

> Inside each async method, we are catching errors when they occur and save them to the reactive property called `error` under the key corresponding to the triggered method:

Reviewer: ```suggestion
Inside each async method, we catch thrown errors and save them in the `error` object under the key corresponding to the called method.
```

```ts
const { addItem, error } = useCart();

> addItem({ product: null }); // triggers an error

Reviewer: ```suggestion
addItem({ product: null }); // throws an error
```

Reviewer: I would avoid using work `trigger` unless there is no alternative

error.value.addItem; // here you have error raised by addItem function
```

:::

> :::details Where can I find an interface of the error property from a certain composable?

Reviewer: ```suggestion
:::details Where can I find an interface of the error property for a specific composable?
```

When you are writing a code inside a script part of the Vue component, your IDE should give you hints on every different type of the composable. That's why you probably do not need to check these interfaces in the core's code.

However, if you still want to check interfaces, you could find them inside [`packages/core/core/src/types.ts`](https://github.com/vuestorefront/vue-storefront/blob/next/packages/core/core/src/types.ts).

> 
> [//]: # 'TODO: This should be added to API reference'
> 

Reviewer: ```suggestion
```

:::

### How to listen for errors?

> Let's imagine you have some global component for error notifications. You want to send information about every new error to this component. But how to know when a new error appears? You can observe an error object with a watcher:

Reviewer: ```suggestion
Let's imagine you have some global component for error notifications. You want to send information about every new error to this component. But how do you know when a new error appears?

You can observe an error object with a watcher:
```

```ts
import { useUiNotification } from '~/composables';

const { cart, error } = useCart();
const { send } = useUiNotification();

watch(() => ({...error.value}), (error, prevError) => {
  if (error.addItem && error.addItem !== prevError.addItem)
    send({ type: 'danger', message: error.addItem.message });
  if (
    error.removeItem &&
    error.removeItem !== prevError.removeItem
  )
    send({ type: 'danger', message: error.removeItem.message });
});
```

In this example, we are using `useUiNotification` - a composable that handles notifications state. You can read more about it in the API reference.

> 
> [//]: # 'TODO: This should be added to API reference'

Reviewer: ```suggestion
```
